llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -0,0 +1,232 @@
+                      << "\n");
+    return Op;
+  }
+
+  return Expander.expandCodeFor(
+      AddRec->evaluateAtIteration(SE.getConstant(APInt(32, 0)), SE),
+      Op->getType(), InsertionPoint);
+}
+
+/// Recursivly check if \p Op and its uses are loop invariant. \return Op or in
+/// the case of a PHINode, the incoming value of the Preheader, once loop
+/// invariance can be determined
+Value *getLoopInvariant(Value *Op, DominatorTree *DT, Loop *CurLoop) {
+  if (auto *PN = dyn_cast<PHINode>(Op)) {
+    LLVM_DEBUG(dbgs() << "Found end: Phi node " << *Op << "\n");
+    if (DT->properlyDominates(PN->getParent(), CurLoop->getHeader()))
+      return Op;
+
+    return PN->getIncomingValueForBlock(CurLoop->getLoopPreheader());
+  }
+
+  // assumption: LICM has already run, so only the occurance of Op within
+  // the CurLoop has to be checked, since it will otherwise be outside the Loop
+  // and loop invariant
+  if (!CurLoop->isLoopInvariant(Op)) {
+    LLVM_DEBUG(dbgs() << "Operand is loop variant " << *Op << "\n");
+    return nullptr;
+  }
+
+  auto *I = dyn_cast<Instruction>(Op);
+  if (!I) {
+    LLVM_DEBUG(dbgs() << "Found end: " << *Op << "\n");
+    return Op;
+  }
+
+  LLVM_DEBUG(dbgs() << "Checking " << *I << "\n");
+  for (Value *Use : I->operands()) {
+    LLVM_DEBUG(dbgs() << "Instruction " << *I << " -> Operand " << *Use
+                      << "\n");
+    if (!getLoopInvariant(Use, DT, CurLoop))
+      return nullptr;
+  }
+  return Op;
+}
+
+// Create the assumption into the Loop PreHeader, that at iteration 0 the
+// condition is true
+void insertMinIterAssumption(ICmpInst &LoopCmpInstr, Loop &CurrentLoop,
+                             ScalarEvolution &SE, AssumptionCache &AC,
+                             DominatorTree *DT) {
+
+  if (!CurrentLoop.getLoopPreheader()) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountAssumptions-Warning: Loop has no "
+                         "preheader, will not insert Assumption!\n");
+    return;
+  }
+
+  Instruction *InsertionPoint = CurrentLoop.getLoopPreheader()->getTerminator();
+  LLVM_DEBUG(dbgs() << "Inserting Assumption before: " << InsertionPoint
+                    << "\n");
+
+  // LoopRotate uses SimplifyQuery to determine, if a Branch is conditional or
+  // not. SimplifyQuery can only take an Assumption into account, if it is
+  // before the to-be-evaluated Compare Instruction. Here they are inserted into
+  // the Preheader, so that the assumption is only valid once and not on every
+  // entry of the loop header.
+  IRBuilder<> Builder(dyn_cast<Instruction>(InsertionPoint));
+
+  SCEVExpander Expander(
+      SE, CurrentLoop.getLoopPreheader()->getModule()->getDataLayout(),
+      "expanded");
+
+  // If AddRecExpr of LHS is available, evaluate at Iteration 0, otherwise
+  // return LHS
+  Value *LHS = expandValueAtZeroIteration(LoopCmpInstr.getOperand(0), SE,
+                                          Expander, &LoopCmpInstr);
+  LHS = getLoopInvariant(LHS, DT, &CurrentLoop);
```

Comment:
See my comment here https://github.com/Xilinx/llvm-aie/pull/227/files#r1869145688

I think it is a bit confusing that `expandValueAtZeroIteration` potentially returns something that isn't the value at iteration 0. This `LHS = getLoopInvariant(LHS, DT, &CurrentLoop);` should be inlined into `expandValueAtZeroIteration`. If the latter cannot be sure that the returned value is exactly the one at iteration 0, it should return `nullptr`.

Reply:
That is a more concise way of looking at `expandValueAtZeroIteration`

Reply:
I updated it accordingly.